src/main/host/syscall/handler/sched.rs

```diff
@@ -0,0 +1,89 @@
+
+use log::warn;
+use nix::errno::Errno;
+use syscall_logger::log_syscall;
+
+// We always report that the thread is running on CPU 0, Node 0
+const CURRENT_CPU: u32 = 0;
+
+const RSEQ_FLAG_UNREGISTER: i32 = 1;
+
+#[repr(C, align(32))]
+#[derive(Debug, Copy, Clone)]
+struct rseq {
+    cpu_id_start: i32,
+    cpu_id: i32,
+    rseq_cs: *mut libc::c_void,
+    flags: u32,
+}
+
+unsafe impl Pod for rseq {}
+
+impl SyscallHandler {
+    #[log_syscall(/* rv */ i32)]
+    pub fn sched_yield(&self, _ctx: &mut ThreadContext, _args: &SysCallArgs) -> SyscallResult {
+        // Do nothing. We already yield and reschedule after some number of
+        // unblocked syscalls.
+        Ok(0.into())
+    }
+
+    #[log_syscall(/* rv */ i32, /* rseq  */*const libc::c_void, /* rseq_len */u32, /* flags */i32, /* sig */u32)]
+    pub fn rseq(&self, ctx: &mut ThreadContext, args: &SysCallArgs) -> SyscallResult {
+        let rseq_ptr = TypedPluginPtr::new::<rseq>(unsafe { args.get(0).as_ptr }.into(), 1);
+        let rseq_len = usize::try_from(unsafe { args.get(1).as_u64 }).unwrap();
+        if rseq_len != std::mem::size_of::<rseq>() {
+            // Probably worth a warning; decent chance that the bug is in Shadow
+            // rather than the calling code.
+            warn!(
+                "rseq_len {} instead of expected {}",
+                rseq_len,
+                std::mem::size_of::<rseq>()
+            );
+            return Err(Errno::EINVAL.into());
+        }
+        let flags = i32::try_from(unsafe { args.get(2).as_i64 }).map_err(|_| Errno::EINVAL)?;
+        let sig = u32::try_from(unsafe { args.get(3).as_u64 }).map_err(|_| Errno::EINVAL)?;
+        self.rseq_impl(ctx, rseq_ptr, flags, sig)
+    }
+
+    fn rseq_impl(
+        &self,
+        ctx: &mut ThreadContext,
+        rseq_ptr: TypedPluginPtr<rseq>,
+        flags: i32,
+        _sig: u32,
+    ) -> SyscallResult {
+        if flags & (!RSEQ_FLAG_UNREGISTER) != 0 {
+            warn!("Unrecognized flags: {}", flags);
+            return Err(Errno::EINVAL.into());
+        }
+        if flags & RSEQ_FLAG_UNREGISTER != 0 {
+            // TODO:
+            // * Validate that an rseq was previously registered
+            // * Validate that `sig` matches registration
+            // * Set the cpu_id of the previously registerd rseq to the uninitialized
+            //   state.
+            return Ok(0.into());
+        }
+        let mem = ctx.process.memory_mut();
+        let mut rseq = mem.memory_ref_mut(rseq_ptr)?;
+
+        // rseq is mostly unimplemented, but also mostly unneeded in Shadow.
+        // We'd only need to implement the "real" functionality if we ever implement
+        // true preemption, in which case we'd need to do something if we ever pre-empted
+        // while the user code was in a restartable sequence. As it is, Shadow only
+        // reschedules threads at system calls, and system calls are disallowed inside
+        // restartable sequences.
```

Comment:
I think the thread can be pre-empted for another reason though, for example in Shadow I think this can happen for signals generated by hardware errors? For example, a critical section dereferences an invalid pointer, the sigsegv handler runs immediately, and then we should jump to the abort instruction once the handler completes.

This probably rarely happens in practice so I don't think we need to handle it, but recovering from signals during a restartable sequence is something Linux does support. Is there a way we can warn on the incorrect behaviour anyways? For example, can we store the `struct rseq` pointer in the `SysCallHandler`, and before running a signal handler read the plugin's `struct rseq_cs` to make sure we're not in a restartable sequence? Is it even straightforward for the shim to know the instruction address before the signal was invoked (is it in one of the signal context objects)?

I'm also fine with not handling this at all, but maybe add a comment that we don't handle this and that weird things might happen.

Reply:
Good point; I was thinking triggering a hardware signal inside an rseq might be undefined behavior the same way making a syscall is, but I'm not really sure.

Conceptually it wouldn't be too difficult to implement more fully, but it's a bit of a rabbit hole right now. I think we'd want to store the `struct rseq` in shared memory in `ShimThreadProtectedSharedMem`, so that the shim-side signal handling code can get to it. We haven't set up bindings for that or any of the other `shadow-shim-helper` yet. I *think* we can probably just rewrite that whole library in Rust and expose back to C rather than investing in safe Rust wrappers, but again it's a bit of a rabbit hole.

Opened https://github.com/shadow/shadow/issues/2139 to track